c10/cuda/CUDACachingAllocator.h

```diff
@@ -246,6 +251,9 @@ class CUDAAllocator : public Allocator {
       size_t alloc_trace_max_entries,
       RecordContext when) = 0;
   virtual void attachOutOfMemoryObserver(OutOfMemoryObserver observer) = 0;
+  // We require AllocatorTraceTracker callback never include Pytorch call to
```

Comment:
//`AllocatorTraceTracker` will be called while the per-device allocator lock is held. Any additional locks taken from within the callback must be proven to always have the lock order <allocator lock> -> <my lock>, otherwise deadlocks will occur. In particular, Python's GIL may be held when calling the allocator so it is unsafe to try to acquire the GIL in this callback.

Reply:
Thank you for the detailed restriction description! I will use it in the comment here.